### PR TITLE
Do not download from external sources

### LINKED ISSUE #3179: AppImage publishing relies on script hosted on another web site
Merged https://github.com/PowerShell/PowerShell/pull/2082 prematurely.  @lzybkr brought to my attention that his concern wasn't addressed.

https://github.com/PowerShell/PowerShell/blob/master/tools/appimage.sh#L41

Relies on

```
https://github.com/probonopd/AppImages/raw/e05cbebc62c86f8c602d74d9050bbfbf10df1c69/functions.sh 
```

which internally gets and executes other scripts.

If any of these scripts are compromised, we'll run arbitrary code as part of our publishing pipeline.
We need to resolve this vulnerability before we start publishing appimages.

CC @raghushantha @andschwa @probonopd

### PR DESCRIPTION
**Pull request by probonopd**
Closes #3179. Be sure to change the line below `NOTE: REPLACE THIS WITH A MIRRORED COPY OF APPIMAGETOOL` as you see fit.

### CODE REVIEW

File: tools/appimage.sh
Comment (SteveL-MSFT) [inline tools/appimage.sh]: Please update URL to https://psgithub.file.core.windows.net/psgithub/appimaged-x86_64.AppImage

**Review by SteveL-MSFT — CHANGES_REQUESTED** [review]

File: tools/appimage.sh
Comment (probonopd) [inline tools/appimage.sh]: @SteveL-MSFT `appimaged` is not `appimagetool`. Please check.

Comment (SteveL-MSFT) [inline tools/appimage.sh]: @probonopd https://psgithub.file.core.windows.net/psgithub/appimagetool-x86_64.AppImage

File: tools/appimage.sh
Comment (SteveL-MSFT) [inline tools/appimage.sh]: Sorry, looks like I used the wrong type of Azure Storage.  Use this instead:

https://psgithub.blob.core.windows.net/files/appimagetool-x86_64.AppImage

**Review by SteveL-MSFT — CHANGES_REQUESTED** [review]

**Comment by SteveL-MSFT** [comment]
Not successful in running appimage.sh to build:

```
PS /home/steve/repos/probonopd/tools> Start-PSPackage                           
VERBOSE: Packaging RID: 'ubuntu.16.04-x64'; Packaging Configuration: 'Linux'
VERBOSE: Packaging Source: 
'/home/steve/repos/probonopd/src/powershell-unix/bin/Linux/netcoreapp1.1/ubuntu
.16.04-x64/publish'
WARNING: -Type was not specified, continuing with deb!
WARNING: Building for Ubuntu 16.04!
WARNING: Suffix not given, building primary PowerShell package!
     roff: /home/steve/repos/probonopd/assets/powershell.1


    Directory: /home/steve/repos/probonopd


Mode                LastWriteTime         Length Name                          
----                -------------         ------ ----                          
------          2/27/17   1:32 PM       31134168 powershell_6.0.0-alpha.9-833-g
                                                 83f5e73-1ubuntu1.16.04.1_amd64
                                                 .deb                          


PS /home/steve/repos/probonopd/tools> ./appimage.sh                             
dpkg-scanpackages: warning: Packages in archive but missing from override file:
dpkg-scanpackages: warning:   powershell
dpkg-scanpackages: info: Wrote 1 entries to output Packages file.
Get:1 file:/home/steve/repos/probonopd/tools/powershell ./ InRelease
Ign:1 file:/home/steve/repos/probonopd/tools/powershell ./ InRelease
Get:2 file:/home/steve/repos/probonopd/tools/powershell ./ Release
Ign:2 file:/home/steve/repos/probonopd/tools/powershell ./ Release
Get:3 file:/home/steve/repos/probonopd/tools/powershell ./ Packages
Ign:3 file:/home/steve/repos/probonopd/tools/powershell ./ Packages
Get:4 file:/home/steve/repos/probonopd/tools/powershell ./ Translation-en_US
Ign:4 file:/home/steve/repos/probonopd/tools/powershell ./ Translation-en_US
Get:5 file:/home/steve/repos/probonopd/tools/powershell ./ Translation-en
Ign:5 file:/home/steve/repos/probonopd/tools/powershell ./ Translation-en
Get:3 file:/home/steve/repos/probonopd/tools/powershell ./ Packages
Ign:3 file:/home/steve/repos/probonopd/tools/powershell ./ Packages
Get:4 file:/home/steve/repos/probonopd/tools/powershell ./ Translation-en_US
Ign:4 file:/home/steve/repos/probonopd/tools/powershell ./ Translation-en_US
Get:5 file:/home/steve/repos/probonopd/tools/powershell ./ Translation-en
Ign:5 file:/home/steve/repos/probonopd/tools/powershell ./ Translation-en
Get:3 file:/home/steve/repos/probonopd/tools/powershell ./ Packages
Ign:3 file:/home/steve/repos/probonopd/tools/powershell ./ Packages
Get:4 file:/home/steve/repos/probonopd/tools/powershell ./ Translation-en_US
Ign:4 file:/home/steve/repos/probonopd/tools/powershell ./ Translation-en_US
Get:5 file:/home/steve/repos/probonopd/tools/powershell ./ Translation-en
Ign:5 file:/home/steve/repos/probonopd/tools/powershell ./ Translation-en
Get:3 file:/home/steve/repos/probonopd/tools/powershell ./ Packages [528 B]
Get:4 file:/home/steve/repos/probonopd/tools/powershell ./ Translation-en_US
Ign:4 file:/home/steve/repos/probonopd/tools/powershell ./ Translation-en_US
Get:5 file:/home/steve/repos/probonopd/tools/powershell ./ Translation-en
Ign:5 file:/home/steve/repos/probonopd/tools/powershell ./ Translation-en
Get:4 file:/home/steve/repos/probonopd/tools/powershell ./ Translation-en_US
Ign:4 file:/home/steve/repos/probonopd/tools/powershell ./ Translation-en_US
Get:5 file:/home/steve/repos/probonopd/tools/powershell ./ Translation-en
Ign:5 file:/home/steve/repos/probonopd/tools/powershell ./ Translation-en
Get:4 file:/home/steve/repos/probonopd/tools/powershell ./ Translation-en_US
Ign:4 file:/home/steve/repos/probonopd/tools/powershell ./ Translation-en_US
Get:5 file:/home/steve/repos/probonopd/tools/powershell ./ Translation-en
Ign:5 file:/home/steve/repos/probonopd/tools/powershell ./ Translation-en
Ign:6 http://archive.ubuntu.com/ubuntu trusty InRelease
Hit:7 http://archive.ubuntu.com/ubuntu trusty Release
Reading package lists... Done 
W: The repository 'file:/home/steve/repos/probonopd/tools/powershell ./ Release' does not have a Release file.
N: Data from such a repository can't be authenticated and is therefore potentially dangerous to use.
N: See apt-secure(8) manpage for repository creation and user configuration details.
  powershell:amd64 Depends on libicu55 [ amd64 ] < none ->  > ( none ) can't be satisfied!
Starting pkgProblemResolver with broken count: 1
Starting 2 pkgProblemResolver with broken count: 1
Investigating (0) powershell [ amd64 ] < none -> 6.0.0-alpha.9-833-g83f5e73-1ubuntu1.16.04.1 > ( shells )
Broken powershell:amd64 Depends on libc6 [ amd64 ] < none -> 2.19-0ubuntu6 > ( libs )
  Considering libc6:amd64 1 as a solution to powershell:amd64 9998
  Re-Instated multiarch-support:amd64
  Re-Instated gcc-4.9-base:amd64
  Re-Instated libgcc1:amd64
  Re-Instated libc6:amd64
Broken powershell:amd64 Depends on libcurl3 [ amd64 ] < none -> 7.35.0-1ubuntu2 > ( libs )
  Considering libcurl3:amd64 1 as a solution to powershell:amd64 9998
  Re-Instated libcomerr2:amd64
  Re-Instated libkrb5support0:amd64
  Re-Instated libk5crypto3:amd64
  Re-Instated libkeyutils1:amd64
  Re-Instated libkrb5-3:amd64
  Re-Instated libgssapi-krb5-2:amd64
  Re-Instated libidn11:amd64
  Re-Instated libgpg-error0:amd64
  Re-Instated libgcrypt11:amd64
  Re-Instated libffi6:amd64
  Re-Instated libp11-kit0:amd64
  Re-Instated libtasn1-6:amd64
  Re-Instated zlib1g:amd64
  Re-Instated libgnutls26:amd64
  Re-Instated libroken18-heimdal:amd64
  Re-Instated libasn1-8-heimdal:amd64
  Re-Instated libhcrypto4-heimdal:amd64
  Re-Instated libheimbase1-heimdal:amd64
  Re-Instated libwind0-heimdal:amd64
  Re-Instated libhx509-5-heimdal:amd64
  Re-Instated libsqlite3-0:amd64
  Re-Instated libkrb5-26-heimdal:amd64
  Re-Instated libheimntlm0-heimdal:amd64
  Re-Instated libgssapi3-heimdal:amd64
  Re-Instated libdb5.3:amd64
  Re-Instated libsasl2-modules-db:amd64
  Re-Instated libsasl2-2:amd64
  Re-Instated libldap-2.4-2:amd64
  Re-Instated librtmp0:amd64
  Re-Instated libbz2-1.0:amd64
  Re-Instated liblzma5:amd64
  Re-Instated libpcre3:amd64
  Re-Instated libselinux1:amd64
  Re-Instated libattr1:amd64
  Re-Instated libacl1:amd64
  Re-Instated tar:amd64
  Re-Instated dpkg:amd64
  Re-Instated perl-base:amd64
  Re-Instated debconf:amd64
  Re-Instated libssl1.0.0:amd64
  Re-Instated libcurl3:amd64
Broken powershell:amd64 Depends on libstdc++6 [ amd64 ] < none -> 4.8.2-19ubuntu1 > ( libs )
  Considering libstdc++6:amd64 1 as a solution to powershell:amd64 9998
  Re-Instated gcc-4.8-base:amd64
  Re-Instated libstdc++6:amd64
Broken powershell:amd64 Depends on libtinfo5 [ amd64 ] < none -> 5.9+20140118-1ubuntu1 > ( libs )
  Considering libtinfo5:amd64 1 as a solution to powershell:amd64 9998
  Re-Instated libtinfo5:amd64
Broken powershell:amd64 Depends on libunwind8 [ amd64 ] < none -> 1.1-2.2ubuntu3 > ( libs )
  Considering libunwind8:amd64 1 as a solution to powershell:amd64 9998
  Re-Instated libunwind8:amd64
Broken powershell:amd64 Depends on libuuid1 [ amd64 ] < none -> 2.20.1-5.1ubuntu20 > ( libs )
  Considering libuuid1:amd64 1 as a solution to powershell:amd64 9998
  Re-Instated libaudit-common:amd64
  Re-Instated libaudit1:amd64
  Re-Instated libpam0g:amd64
  Re-Instated libsemanage-common:amd64
  Re-Instated libsepol1:amd64
  Re-Instated libustr-1.0-1:amd64
  Re-Instated libsemanage1:amd64
  Re-Instated libpam-modules-bin:amd64
  Re-Instated libpam-modules:amd64
  Re-Instated sensible-utils:amd64
  Re-Instated debianutils:amd64
  Re-Instated passwd:amd64
  Re-Instated libuuid1:amd64
Broken powershell:amd64 Depends on libicu55 [ amd64 ] < none ->  > ( none )
Done
E: Unable to correct problems, you have held broken packages.
wget: missing URL
Usage: wget [OPTION]... [URL]...

Try `wget --help' for more options.
cp: cannot stat '../../assets/Powershell_256.png': No such file or directory
cp: cannot stat '../../assets/AppImageThirdPartyNotices.txt': No such file or directory
'./lib64/' -> './usr/./lib64'
mv: cannot stat './usr/lib/x86_64-linux-gnu/*': No such file or directory
mv: cannot stat './usr/lib/pulseaudio/*.so': No such file or directory
libc.so.6 libdl.so.2 libstdc++.so.6 libm.so.6 libpthread.so.0 libresolv.so.2 libGL.so.1 libdrm.so.2 libxcb.so.1 libX11.so.6 libgio-2.0.so.0 libgdk-x11-2.0.so.0 libgtk-x11-2.0.so.0 libasound.so.2 libgdk_pixbuf-2.0.so.0 libfontconfig.so.1 libselinux.so.1 libcom_err.so.2 libcrypt.so.1 libexpat.so.1 libgcc_s.so.1 libglib-2.0.so.0 libgpg-error.so.0 libgssapi_krb5.so.2 libhcrypto.so.4 libhx509.so.5 libICE.so.6 libidn.so.11 libk5crypto.so.3 libkeyutils.so.1 libkrb5.so.26 libkrb5.so.3 libkrb5support.so.0 libm.so.6 libp11-kit.so.0 libpthread.so.0 libresolv.so.2 libroken.so.18 librt.so.1 libSM.so.6 libusb-1.0.so.0 libuuid.so.1 libwind.so.0 libz.so.1 libGL.so.1 libdrm.so.2 libgobject-2.0.so.0 libgpg-error.so.0 libxcb.so.1
rm: cannot remove 'usr/lib/x86_64-linux-gnu/': No such file or directory
6.0.0-alpha.9
--2017-02-27 13:36:18--  https://psgithub.blob.core.windows.net/files/appimagetool-x86_64.AppImage
Resolving psgithub.blob.core.windows.net (psgithub.blob.core.windows.net)... 52.239.160.100
Connecting to psgithub.blob.core.windows.net (psgithub.blob.core.windows.net)|52.239.160.100|:443... connected.
HTTP request sent, awaiting response... 200 OK

    The file is already fully retrieved; nothing to do.

Arch: 86_64
DESTINATION not specified, so assuming PowerShell-x86_64.AppImage
/home/steve/repos/probonopd/tools/powershell/powershell.AppDir should be packaged as PowerShell-x86_64.AppImage
powershell{.png,.svg,.svgz,.xpm} not present but defined in desktop file
cp: cannot stat './PowerShell*AppImage': No such file or directory
```

**Comment by probonopd** [comment]
The `WARNING: Building for Ubuntu 16.04!` looks suspicious, can you build for 14.04?

**Comment by SteveL-MSFT** [comment]
@probonopd sorry was in meetings rest of day and then we got hit by snow over here, working on this today

**Comment by SteveL-MSFT** [comment]
@probonopd still not working on ubuntu 14.04

```
cp: cannot stat ‘../powershell_*_amd64.deb’: No such file or directory
./appimage.sh: line 416: dpkg-scanpackages: command not found
Ign file: ./ InRelease
Ign file: ./ Release.gpg
Ign file: ./ Release
Ign file: ./ Translation-en_US        
Ign file: ./ Translation-en           
Ign http://archive.ubuntu.com trusty InRelease            
Hit http://archive.ubuntu.com trusty Release.gpg
Hit http://archive.ubuntu.com trusty Release
Hit http://archive.ubuntu.com trusty/main amd64 Packages
Hit http://archive.ubuntu.com trusty/universe amd64 Packages
Hit http://archive.ubuntu.com trusty/main i386 Packages
Hit http://archive.ubuntu.com trusty/universe i386 Packages
Hit http://archive.ubuntu.com trusty/main Translation-en
Hit http://archive.ubuntu.com trusty/universe Translation-en
Ign http://archive.ubuntu.com trusty/main Translation-en_US
Ign http://archive.ubuntu.com trusty/universe Translation-en_US
Reading package lists... Done 
E: Unable to locate package powershell
wget: missing URL
Usage: wget [OPTION]... [URL]...

Try `wget --help' for more options.
find: `../*.deb': No such file or directory
rm: cannot remove ‘usr/bin/powershell’: No such file or directory
mv: cannot stat ‘opt/microsoft/powershell/*/*’: No such file or directory
cp: cannot stat ‘../../assets/Powershell_256.png’: No such file or directory
cp: cannot stat ‘../../assets/AppImageThirdPartyNotices.txt’: No such file or directory
‘./lib64/’ -> ‘./usr/./lib64’
mv: cannot stat ‘./usr/lib/x86_64-linux-gnu/*’: No such file or directory
mv: cannot stat ‘./usr/lib/pulseaudio/*.so’: No such file or directory
mv: cannot stat ‘usr/local/share/man’: No such file or directory
libc.so.6 libdl.so.2 libstdc++.so.6 libm.so.6 libpthread.so.0 libresolv.so.2 libGL.so.1 libdrm.so.2 libxcb.so.1 libX11.so.6 libgio-2.0.so.0 libgdk-x11-2.0.so.0 libgtk-x11-2.0.so.0 libasound.so.2 libgdk_pixbuf-2.0.so.0 libfontconfig.so.1 libselinux.so.1 libcom_err.so.2 libcrypt.so.1 libexpat.so.1 libgcc_s.so.1 libglib-2.0.so.0 libgpg-error.so.0 libgssapi_krb5.so.2 libhcrypto.so.4 libhx509.so.5 libICE.so.6 libidn.so.11 libk5crypto.so.3 libkeyutils.so.1 libkrb5.so.26 libkrb5.so.3 libkrb5support.so.0 libm.so.6 libp11-kit.so.0 libpthread.so.0 libresolv.so.2 libroken.so.18 librt.so.1 libSM.so.6 libusb-1.0.so.0 libuuid.so.1 libwind.so.0 libz.so.1 libGL.so.1 libdrm.so.2 libgobject-2.0.so.0 libgpg-error.so.0 libxcb.so.1
rm: cannot remove ‘opt/’: No such file or directory
rm: cannot remove ‘usr/lib/x86_64-linux-gnu/’: No such file or directory
rm: cannot remove ‘usr/share/’: No such file or directory
find: `../*.deb': No such file or directory

./appimage.sh: line 76: ./usr/bin/powershell.wrapper: No such file or directory
chmod: cannot access ‘./usr/bin/powershell.wrapper’: No such file or directory
sed: can't read usr/bin/powershell.wrapper: No such file or directory
--2017-02-28 10:46:52--  https://psgithub.blob.core.windows.net/files/appimagetool-x86_64.AppImage
Resolving psgithub.blob.core.windows.net (psgithub.blob.core.windows.net)... 52.239.160.100
Connecting to psgithub.blob.core.windows.net (psgithub.blob.core.windows.net)|52.239.160.100|:443... connected.
HTTP request sent, awaiting response... 200 OK

    The file is already fully retrieved; nothing to do.

WARNING: appstreamcli is missing, please install it if you want to use AppStream metadata
WARNING: gpg2 is missing, please install it if you want to create digital signatures
Arch: 86_64
DESTINATION not specified, so assuming PowerShell-x86_64.AppImage
/home/steve/repos/probonopd/tools/powershell/powershell.AppDir should be packaged as PowerShell-x86_64.AppImage
powershell{.png,.svg,.svgz,.xpm} not present but defined in desktop file
cp: cannot stat ‘./PowerShell*AppImage’: No such file or directory
```

**Comment by probonopd** [comment]
This time `cp: cannot stat ‘../powershell_*_amd64.deb’: No such file or directory` means that you are missing the PowerShell deb to begin with...

**Review by SteveL-MSFT — APPROVED** [review]

**Comment by SteveL-MSFT** [comment]
Ok, got it working by running tool/appimage.sh from root of repo folder rather than inside the tools folder.   Tested the AppImage and worked fine on 14.04.  Thanks @probonopd

**Comment by SteveL-MSFT** [comment]
Worked on 16.04 after running script from root of repo.

**Comment by SteveL-MSFT** [comment]
@andschwa anything else you need before we merge this?

**Comment by andschwa** [comment]
@SteveL-MSFT not that I know of!

**Comment by SteveL-MSFT** [comment]
@andschwa since you're the maintainer assigned to this, can you merge :)

**Pull request merged**
